src/storage/src/source/generator/ldbc.rs

```diff
@@ -0,0 +1,247 @@
+    record: StringRecord,
+    arena: RowArena,
+    offset: MzOffset,
+    pending: Option<(usize, Row)>,
+    progress_outputs: Vec<usize>,
+}
+
+impl Context {
+    /// Opens the next file if `self.rdr` is `None`. `self.rdr` is set to `None` if there are no
+    /// more files to read.
+    fn next_file(&mut self) {
+        while self.rdr.is_none() {
+            let Some((mut path, mut entry)) = self.entries.next() else {
+                return;
+            };
+            // Gunzip gz files, removing the .gz extension.
+            if matches!(path.extension().and_then(|s| s.to_str()), Some("gz")) {
+                let mut buf = Vec::new();
+                if let Err(err) = GzDecoder::new(entry.as_slice()).read_to_end(&mut buf) {
+                    error!("error gunzipping inner ldbc file {:?}: {}", path, err);
+                    continue;
+                }
+                path = path.with_extension("");
+                entry = buf;
+            }
+            // Skip non-CSV files.
+            if !matches!(path.extension().and_then(|s| s.to_str()), Some("csv")) {
+                continue;
+            }
+            let parts = path.parent().and_then(|path| {
+                let mut parts = path.components().rev();
+                let Some(entity) = parts.next() else {
+                    return None;
+                };
+                let Some(entity) = entity.as_os_str().to_str() else {
+                    return None;
+                };
+                let Some(load) = parts.next() else {
+                    return None;
+                };
+                let Some(load) = load.as_os_str().to_str() else {
+                    return None;
+                };
+                Some((load, entity))
+            });
+            let (output, desc) = match parts {
+                Some(("static" | "dynamic", entity)) => {
+                    let entity = entity.to_lowercase();
+                    let Some((output, desc)) = LDBC_DESCS.get(entity.as_str()) else {
+                        error!("could not find ldbc relation: {entity}");
+                        continue;
+                    };
+                    (output, desc)
+                }
+                Some(_) => continue,
+                None => continue,
+            };
+            let reader = Bytes::from(entry).reader();
+            let rdr = ReaderBuilder::new().delimiter(b'|').from_reader(reader);
+            self.rdr = Some(CsvFile {
+                rdr,
+                desc,
+                output: *output,
+            });
+        }
+    }
+
+    /// Reads the next record out of `self.rdr`. Returns `None` if `self.rdr` is out of records.
+    fn next_record(&mut self) -> Option<(usize, Row)> {
+        self.rdr
+            .as_mut()
+            .and_then(|csv| match csv.rdr.read_record(&mut self.record) {
+                Ok(true) => {
+                    let row = Row::pack(self.record.iter().zip(csv.desc.iter_types()).map(
```

Comment:
`zip` will early return if the iterators do not have equal lengths which means we can end up persisting invalid `Row`s which may blow up later in dataflows. We should verify that we get the expected number of datums and produce an error if not